host_core/lib/host_core/application.ex

```diff
@@ -26,4 +32,67 @@ defmodule HostCore.Application do
     Logger.info("Starting Host Core")
     Supervisor.start_link(children, opts)
   end
+
+  defp retrieve_nats_env() do
+    rpc_host =
+      case System.get_env("WASMCLOUD_RPC_HOST") do
+        nil -> '0.0.0.0'
+        host -> host
+      end
+
+    rpc_port =
+      case System.get_env("WASMCLOUD_RPC_PORT") do
+        nil -> 4222
+        port -> String.to_integer(port)
+      end
+
+    ctl_host =
+      case System.get_env("WASMCLOUD_CTL_HOST") do
+        nil -> '0.0.0.0'
+        host -> host
+      end
+
+    ctl_port =
+      case System.get_env("WASMCLOUD_CTL_PORT") do
+        nil -> 4222
+        port -> String.to_integer(port)
+      end
+
+    {rpc_host, rpc_port, ctl_host, ctl_port}
+  end
+
+  defp start_gnat({rpc_host, rpc_port, ctl_host, ctl_port}) do
+    :ok = lattice_connect(rpc_host, rpc_port)
+    :ok = control_connect(ctl_host, ctl_port)
+  end
+
+  defp lattice_connect(rpc_host, rpc_port) do
+    case Gnat.start_link(%{host: rpc_host, port: rpc_port}, name: :lattice_nats) do
+      {:ok, _gnat} ->
+        :ok
+
+      {:error, :econnrefused} ->
+        Logger.warn(
+          "Failed to connect to lattice at #{rpc_host}:#{rpc_port}, retrying in 1 second..."
+        )
+
+        Process.sleep(1000)
+        lattice_connect(rpc_host, rpc_port)
+    end
+  end
+
+  defp control_connect(ctl_host, ctl_port) do
+    case Gnat.start_link(%{host: ctl_host, port: ctl_port}, name: :control_nats) do
+      {:ok, _gnat} ->
+        :ok
+
+      {:error, :econnrefused} ->
+        Logger.warn(
+          "Failed to connect to control interface at #{ctl_host}:#{ctl_port}, retrying in 1 second..."
+        )
+
+        Process.sleep(1000)
```

Comment:
Same comment as above, you can do this in a `send_after` loop.

For bonus points, see about making the reconnect attempts do exponential backoff ;)